In [1]:
from sqlalchemy import create_engine
from sqlalchemy import inspect
import pandas as pd

## Querying from that database into Python

#### 1. Connect Python to mta database

In [2]:
engine = create_engine("sqlite:///mta_data.db")
insp = inspect(engine)
print(insp.get_table_names())

['mta_data']


#### 2. Explore the Database
   - Querying from that database into Python via SQLAlchemy is required

In [3]:
mta_data = pd.read_sql('SELECt * FROM mta_data order by DATE ASC limit 10;', engine)
mta_data

CA  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  04/24/2021  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  04/24/2021  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  04/24/2021  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  04/24/2021  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  04/24/2021  16:00:00   
5  A002  R051  02-00-00   59 ST  NQR456W      BMT  04/24/2021  20:00:00   
6  A002  R051  02-00-01   59 ST  NQR456W      BMT  04/24/2021  00:00:00   
7  A002  R051  02-00-01   59 ST  NQR456W      BMT  04/24/2021  04:00:00   
8  A002  R051  02-00-01   59 ST  NQR456W      BMT  04/24/2021  08:00:00   
9  A002  R051  02-00-01   59 ST  NQR456W      BMT  04/24/2021  12:00:00   

      DESC  ENTRIES    EXITS  
0  REGULAR  7561288  2582898  
1  REGULAR  7561295  2582903  
2  REGULAR  7561302  2582918  
3  REGULAR  7561336  2582977  
4  REGULAR  7561426  2583008  
5  REGULAR  7561578  2583031  
6  REGULAR  6698358  1523748  
7  REGULAR  6698362  1523749  
8  REGULAR  6698370  1523761  
9  REGULAR  6698405  1523790

In [6]:
# check if the data is collected within 
mta_data = pd.read_sql('SELECt COUNT(DISTINCT DATE) FROM mta_data order by DATE ASC;', engine)
mta_data

COUNT(DISTINCT DATE)
0                   140

In [8]:
mta_data = pd.read_sql('select count(*) from mta_data', engine)
mta_data

count(*)
0   4187239

In [9]:
# update the C/A column name to CA : easier to handle 
conn =  engine.connect()
conn.execute('ALTER TABLE mta_data RENAME COLUMN "C/A" TO "CA"')

In [10]:
mta_data_entries = pd.read_sql('SELECT * FROM mta_data order by ENTRIES desc limit 10', engine)
mta_data_entries


CA  UNIT       SCP   STATION LINENAME DIVISION        DATE      TIME  \
0  J003  R352  00-00-00  HEWES ST       JM      BMT  04/24/2021  01:00:00   
1  J003  R352  00-00-00  HEWES ST       JM      BMT  04/24/2021  05:00:00   
2  J003  R352  00-00-00  HEWES ST       JM      BMT  04/24/2021  09:00:00   
3  J003  R352  00-00-00  HEWES ST       JM      BMT  04/24/2021  13:00:00   
4  J003  R352  00-00-00  HEWES ST       JM      BMT  04/24/2021  17:00:00   
5  J003  R352  00-00-00  HEWES ST       JM      BMT  04/24/2021  21:00:00   
6  J003  R352  00-00-00  HEWES ST       JM      BMT  04/25/2021  01:00:00   
7  J003  R352  00-00-00  HEWES ST       JM      BMT  04/25/2021  05:00:00   
8  J003  R352  00-00-00  HEWES ST       JM      BMT  04/25/2021  09:00:00   
9  J003  R352  00-00-00  HEWES ST       JM      BMT  04/25/2021  13:00:00   

      DESC     ENTRIES      EXITS  
0  REGULAR  2147429816  671147622  
1  REGULAR  2147429813  671147623  
2  REGULAR  2147429789  671147631  
3  REGULAR  2147429724  671147646  
4  REGULAR  2147429654  671147677  
5  REGULAR  2147429588  671147700  
6  REGULAR  2147429567  671147707  
7  REGULAR  2147429563  671147714  
8  REGULAR  2147429544  671147731  
9  REGULAR  2147429516  671147757

In [91]:
mta_data_exits = pd.read_sql('SELECT * FROM mta_data order by EXITS desc limit 10', engine)
mta_data_exits

CA  UNIT       SCP         STATION LINENAME DIVISION        DATE  \
0  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/24/2021   
1  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/24/2021   
2  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/24/2021   
3  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/24/2021   
4  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/24/2021   
5  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/24/2021   
6  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/25/2021   
7  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/25/2021   
8  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/25/2021   
9  R523  R147  00-00-04  61 ST WOODSIDE        7      IRT  04/25/2021   

       TIME     DESC     ENTRIES       EXITS  
0  01:00:00  REGULAR  1843804931  2122941295  
1  05:00:00  REGULAR  1843804929  2122941276  
2  09:00:00  REGULAR  1843804819  2122941186  
3  13:00:00  REGULAR  1843804658  2122940992  
4  17:00:00  REGULAR  1843804519  2122940688  
5  21:00:00  REGULAR  1843804443  2122940327  
6  01:00:00  REGULAR  1843804424  2122940153  
7  05:00:00  REGULAR  1843804418  2122940134  
8  09:00:00  REGULAR  1843804369  2122940070  
9  13:00:00  REGULAR  1843804260  2122939944

In [69]:
mta_data = pd.read_sql('SELECT * FROM mta_data order by ENTRIES asc limit 10', engine)
mta_data

CA  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/21/2021  00:00:00   
1  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/21/2021  04:00:00   
2  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/21/2021  08:00:00   
3  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/21/2021  12:00:00   
4  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/21/2021  16:00:00   
5  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/21/2021  20:00:00   
6  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/22/2021  00:00:00   
7  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/22/2021  04:00:00   
8  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/22/2021  08:00:00   
9  A002  R051  02-05-01   59 ST  NQR456W      BMT  08/22/2021  12:00:00   

      DESC  ENTRIES  EXITS  
0  REGULAR        0    910  
1  REGULAR        0    910  
2  REGULAR        0    910  
3  REGULAR        0    910  
4  REGULAR        0    910  
5  REGULAR        0    910  
6  REGULAR        0    910  
7  REGULAR        0    910  
8  REGULAR        0    910  
9  REGULAR        0    910

the numbers in ENTRIES and EXITS doesn't make sense : 
Zero in entries but 910 in exits & 0 in EXITS but 524476 in ENTRIES


In [71]:
mta_data = pd.read_sql('SELECT MAX(ENTRIES), MIN(ENTRIES), AVG(ENTRIES), MAX(EXITS), MIN(EXITS), AVG(EXITS) FROM mta_data order by ENTRIES asc limit 10', engine)
mta_data

MAX(ENTRIES)  MIN(ENTRIES)  AVG(ENTRIES)  MAX(EXITS)  MIN(EXITS)  \
0    2147429816             0  4.176588e+07  2122941295           0   

     AVG(EXITS)  
0  3.341785e+07

In [152]:
# check total number of turnstiles each date
mta_data = pd.read_sql("SELECT *, count(*) FROM mta_data group by CA, UNIT, SCP, DATE order by count(*) desc;", engine)
mta_data.head(20)

CA  UNIT       SCP          STATION     LINENAME DIVISION        DATE  \
0   N110  R283  00-00-00     LAFAYETTE AV            C      IND  07/02/2021   
1   N110  R283  00-00-01     LAFAYETTE AV            C      IND  07/02/2021   
2   N110  R283  00-00-02     LAFAYETTE AV            C      IND  07/02/2021   
3   N110  R283  00-03-00     LAFAYETTE AV            C      IND  07/02/2021   
4   N110  R283  00-03-01     LAFAYETTE AV            C      IND  07/02/2021   
5   N110  R283  00-03-02     LAFAYETTE AV            C      IND  07/02/2021   
6   N110  R283  00-05-00     LAFAYETTE AV            C      IND  07/02/2021   
7   N110  R283  00-05-01     LAFAYETTE AV            C      IND  07/02/2021   
8   N110  R283  00-06-00     LAFAYETTE AV            C      IND  07/02/2021   
9   N110  R283  00-06-01     LAFAYETTE AV            C      IND  07/02/2021   
10  E003  R369  00-00-00  FT HAMILTON PKY            D      BMT  05/04/2021   
11  E003  R369  00-00-01  FT HAMILTON PKY            D      BMT  05/04/2021   
12  E003  R369  00-00-02  FT HAMILTON PKY            D      BMT  05/04/2021   
13  E003  R369  00-03-00  FT HAMILTON PKY            D      BMT  05/04/2021   
14  E003  R369  00-03-01  FT HAMILTON PKY            D      BMT  05/04/2021   
15  E003  R369  00-05-00  FT HAMILTON PKY            D      BMT  05/04/2021   
16  E003  R369  00-05-01  FT HAMILTON PKY            D      BMT  05/04/2021   
17  R147  R033  04-00-00   TIMES SQ-42 ST  1237ACENQRS      IRT  07/03/2021   
18  R147  R033  04-00-01   TIMES SQ-42 ST  1237ACENQRS      IRT  07/03/2021   
19  R147  R033  04-00-02   TIMES SQ-42 ST  1237ACENQRS      IRT  07/03/2021   

        TIME        DESC     ENTRIES      EXITS  count(*)  
0   01:00:00     REGULAR       19045      25148        55  
1   01:00:00     REGULAR        5308       8176        55  
2   01:00:00     REGULAR     1176925    2248870        55  
3   01:00:00     REGULAR      502859    1771856        55  
4   01:00:00     REGULAR     1057467     810958        55  
5   01:00:00     REGULAR      524986     568114        55  
6   01:00:00     REGULAR     2566975    1591492        55  
7   01:00:00     REGULAR     5098584    1193233        55  
8   01:00:00     REGULAR  1962578533  318841149        55  
9   01:00:00     REGULAR     8442813     912706        55  
10  01:00:00  RECOVR AUD     2210868    1001190        38  
11  01:00:00  RECOVR AUD     2007132    2344251        38  
12  01:00:00  RECOVR AUD     2876331    1537153        38  
13  01:00:00  RECOVR AUD      767424     838818        38  
14  01:00:00  RECOVR AUD      137333     130104        38  
15  01:00:00  RECOVR AUD        8862      15132        38  
16  01:00:00  RECOVR AUD     1424816    1994299        38  
17  00:00:00     REGULAR    18847741   15950847        31  
18  00:00:00     REGULAR      167855      83180        31  
19  00:00:00     REGULAR    21298379   12393508        31

In [154]:
mta_data = pd.read_sql("select * from mta_data where CA='E003' AND UNIT='R369' AND SCP='00-05-01' AND DATE='05/04/2021' order by DATE, TIME;", engine)
mta_data


CA  UNIT       SCP          STATION LINENAME DIVISION        DATE  \
0   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
1   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
2   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
3   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
4   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
5   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
6   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
7   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
8   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
9   E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
10  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
11  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
12  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
13  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
14  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
15  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
16  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
17  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
18  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
19  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
20  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
21  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
22  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
23  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
24  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
25  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
26  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
27  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
28  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
29  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
30  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
31  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
32  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
33  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
34  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
35  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
36  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   
37  E003  R369  00-05-01  FT HAMILTON PKY        D      BMT  05/04/2021   

        TIME        DESC  ENTRIES    EXITS  
0   01:00:00  RECOVR AUD  1424816  1994299  
1   05:00:00  RECOVR AUD  1424817  1994302  
2   09:00:00  RECOVR AUD  1424904  1994357  
3   13:00:00  RECOVR AUD  1424933  1994424  
4   13:04:27     REGULAR  1424933  1994426  
5   13:05:49     REGULAR  1424933  1994426  
6   13:07:07     REGULAR  1424933  1994426  
7   13:08:28     REGULAR  1424933  1994426  
8   13:09:48     REGULAR  1424933  1994426  
9   13:11:08     REGULAR  1424933  1994426  
10  13:12:28     REGULAR  1424933  1994429  
11  13:13:48     REGULAR  1424934  1994429  
12  13:15:08     REGULAR  1424934  1994429  
13  13:16:28     REGULAR  1424935  1994429  
14  13:17:48     REGULAR  1424935  1994429  
15  13:19:08     REGULAR  1424935  1994429  
16  13:20:28     REGULAR  1424935  1994429  
17  13:21:48     REGULAR  1424935  1994429  
18  13:23:08     REGULAR  1424936  1994429  
19  13:24:28     REGULAR  1424936  1994429  
20  13:25:49     REGULAR  1424936  1994429  
21  13:27:09     REGULAR  1424936  1994430  
22  13:28:29     REGULAR  1424936  1994432  


In [155]:
# check total number of turnstiles each date
mta_data = pd.read_sql("SELECT *, count(*) FROM mta_data group by CA, UNIT, SCP, DATE order by count(*) asc;", engine)
mta_data.head(20)

CA  UNIT       SCP          STATION LINENAME DIVISION        DATE  \
0    A047  R087  00-06-01        CITY HALL      NRW      BMT  05/26/2021   
1    A058  R001  01-00-00  WHITEHALL S-FRY      R1W      BMT  07/09/2021   
2    A058  R001  01-00-00  WHITEHALL S-FRY      R1W      BMT  07/20/2021   
3    A069  R044  01-03-02      CHAMBERS ST    JZ456      BMT  05/29/2021   
4    B020  R263  00-06-02         AVENUE H       BQ      BMT  06/15/2021   
5    B021  R228  00-03-01         AVENUE J       BQ      BMT  08/18/2021   
6    B025  R150  00-00-01         AVENUE U       BQ      BMT  07/22/2021   
7    G001  R151  00-05-00  CONEY IS-STILLW     DFNQ      BMT  06/20/2021   
8   H007A  R248  02-05-00             1 AV        L      BMT  06/16/2021   
9   H007A  R248  02-05-00             1 AV        L      BMT  06/17/2021   
10  H007A  R248  02-05-00             1 AV        L      BMT  06/22/2021   
11   H041  R152  00-05-01  CANARSIE-ROCKAW        L      BMT  08/26/2021   
12   J001  R460  01-00-00         MARCY AV      JMZ      BMT  05/04/2021   
13   J002  R460  00-06-00         MARCY AV      JMZ      BMT  04/29/2021   
14   J003  R352  00-06-00         HEWES ST       JM      BMT  06/30/2021   
15   J003  R352  00-06-00         HEWES ST       JM      BMT  07/09/2021   
16   J037  R009  00-06-01           121 ST       JZ      BMT  07/27/2021   
17  JFK02  R535  01-00-02   HOWARD BCH JFK        A      IND  07/09/2021   
18   K026  R100  00-05-01  METROPOLITAN AV        M      BMT  07/29/2021   
19   N019  R101  01-00-00           145 ST     ABCD      IND  07/16/2021   

        TIME        DESC     ENTRIES      EXITS  count(*)  
0   20:00:00     REGULAR      262161          7         1  
1   09:33:44     REGULAR     1951194    4628405         1  
2   17:00:00     REGULAR     1951212    4640239         1  
3   20:00:00     REGULAR          14         15         1  
4   12:00:00     REGULAR         544          0         1  
5   00:00:00     REGULAR      120755     151546         1  
6   01:00:00     REGULAR     8049630    2357391         1  
7   21:00:00     REGULAR      393219          0         1  
8   00:00:00  RECOVR AUD      458754          0         1  
9   20:00:00     REGULAR      458758          0         1  
10  16:00:00  RECOVR AUD          23          0         1  
11  05:00:00     REGULAR           0        155         1  
12  01:00:00     REGULAR     8801164    1240853         1  
13  01:00:00     REGULAR      393152     437252         1  
14  01:00:00     REGULAR       84684     604534         1  
15  09:36:48     REGULAR       84688     608721         1  
16  08:29:53     REGULAR  1946307906  520339739         1  
17  01:00:00     REGULAR      187136     121885         1  
18  20:00:00     REGULAR          68         62         1  
19  00:00:00     REGULAR     7775646   17208917         1

In [137]:
mta_data = pd.read_sql("select * from mta_data where CA='N019' AND UNIT='R101' AND SCP='01-00-00' AND DATE='07/16/2021';", engine)
mta_data

CA  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  N019  R101  01-00-00  145 ST     ABCD      IND  07/16/2021  00:00:00   

      DESC  ENTRIES     EXITS  
0  REGULAR  7775646  17208917

In [131]:
mta_data.tail(20)

CA  UNIT       SCP        DATE  count(*)
629199  R147  R033  04-00-06  07/03/2021        31
629200  R147  R033  04-05-00  07/03/2021        31
629201  R147  R033  04-05-01  07/03/2021        31
629202  E003  R369  00-00-00  05/04/2021        38
629203  E003  R369  00-00-01  05/04/2021        38
629204  E003  R369  00-00-02  05/04/2021        38
629205  E003  R369  00-03-00  05/04/2021        38
629206  E003  R369  00-03-01  05/04/2021        38
629207  E003  R369  00-05-00  05/04/2021        38
629208  E003  R369  00-05-01  05/04/2021        38
629209  N110  R283  00-00-00  07/02/2021        55
629210  N110  R283  00-00-01  07/02/2021        55
629211  N110  R283  00-00-02  07/02/2021        55
629212  N110  R283  00-03-00  07/02/2021        55
629213  N110  R283  00-03-01  07/02/2021        55
629214  N110  R283  00-03-02  07/02/2021        55
629215  N110  R283  00-05-00  07/02/2021        55
629216  N110  R283  00-05-01  07/02/2021        55
629217  N110  R283  00-06-00  07/02/2021        55
629218  N110  R283  00-06-01  07/02/2021        55

In [88]:
mta_data = pd.read_sql('SELECT  CA, UNIT, SCP, DATE, TIME, SUM(ENTRIES), SUM(EXITS) FROM mta_data group by CA, UNIT, SCP, DATE order by DATE, TIME limit 20;', engine)
mta_data

CA  UNIT       SCP        DATE      TIME  SUM(ENTRIES)  SUM(EXITS)
0   A002  R051  02-00-00  04/24/2021  00:00:00      45368225    15497735
1   A002  R051  02-00-01  04/24/2021  00:00:00      40190581     9142700
2   A002  R051  02-03-00  04/24/2021  00:00:00       8556239    32311146
3   A002  R051  02-03-01  04/24/2021  00:00:00       9486609    14790482
4   A002  R051  02-03-02  04/24/2021  00:00:00         80547      103998
5   A002  R051  02-03-03  04/24/2021  00:00:00      36849104    41951849
6   A002  R051  02-03-04  04/24/2021  00:00:00      45362901    23420036
7   A002  R051  02-03-05  04/24/2021  00:00:00      72842586     8783956
8   A002  R051  02-03-06  04/24/2021  00:00:00      56793335     3089100
9   A002  R051  02-05-00  04/24/2021  00:00:00       3146574           0
10  A002  R051  02-05-01  04/24/2021  00:00:00             0        5418
11  A002  R051  02-06-00  04/24/2021  00:00:00        769783      111047
12  A010  R080  00-00-00  04/24/2021  00:00:00      30677914    18051865
13  A010  R080  00-00-01  04/24/2021  00:00:00      94060552    37364112
14  A010  R080  00-00-02  04/24/2021  00:00:00      68766313    27794751
15  A010  R080  00-00-03  04/24/2021  00:00:00      23249695    11501159
16  A010  R080  00-00-04  04/24/2021  00:00:00      29037626    14114968
17  A010  R080  00-00-05  04/24/2021  00:00:00      13812612     4015058
18  A010  R080  00-00-06  04/24/2021  00:00:00     119434238    33163513
19  A010  R080  00-00-07  04/24/2021  00:00:00      23804801     3704307

In [80]:
mta_data.head(10)

CA  UNIT       SCP        DATE      TIME  SUM(ENTRIES)  SUM(EXITS)
0  A002  R051  02-00-00  04/24/2021  00:00:00      45368225    15497735
1  A002  R051  02-00-01  04/24/2021  00:00:00      40190581     9142700
2  A002  R051  02-03-00  04/24/2021  00:00:00       8556239    32311146
3  A002  R051  02-03-01  04/24/2021  00:00:00       9486609    14790482
4  A002  R051  02-03-02  04/24/2021  00:00:00         80547      103998
5  A002  R051  02-03-03  04/24/2021  00:00:00      36849104    41951849
6  A002  R051  02-03-04  04/24/2021  00:00:00      45362901    23420036
7  A002  R051  02-03-05  04/24/2021  00:00:00      72842586     8783956
8  A002  R051  02-03-06  04/24/2021  00:00:00      56793335     3089100
9  A002  R051  02-05-00  04/24/2021  00:00:00       3146574           0

In [11]:
mta_data=pd.read_sql("select * from mta_data;", engine)
mta_data.head(10)

CA  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/04/2021  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/04/2021  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/04/2021  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/04/2021  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/04/2021  16:00:00   
5  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/04/2021  20:00:00   
6  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/05/2021  00:00:00   
7  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/05/2021  04:00:00   
8  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/05/2021  08:00:00   
9  A002  R051  02-00-00   59 ST  NQR456W      BMT  09/05/2021  12:00:00   

      DESC  ENTRIES    EXITS  
0  REGULAR  7629452  2610530  
1  REGULAR  7629460  2610532  
2  REGULAR  7629470  2610549  
3  REGULAR  7629521  2610587  
4  REGULAR  7629635  2610604  
5  REGULAR  7629817  2610625  
6  REGULAR  7629865  2610638  
7  REGULAR  7629870  2610644  
8  REGULAR  7629879  2610654  
9  REGULAR  7629917  2610679

In [4]:
mta_data=pd.read_sql("select * from mta_data where DATE BETWEEN '05-01-2021' AND '09-01-2021';", engine)
mta_data.head(10)

CA  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  16:00:00   
5  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  20:00:00   
6  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/29/2021  00:00:00   
7  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/29/2021  04:00:00   
8  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/29/2021  08:00:00   
9  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/29/2021  12:00:00   

      DESC  ENTRIES    EXITS  
0  REGULAR  7625998  2609113  
1  REGULAR  7626014  2609117  
2  REGULAR  7626024  2609135  
3  REGULAR  7626086  2609191  
4  REGULAR  7626206  2609221  
5  REGULAR  7626348  2609238  
6  REGULAR  7626446  2609254  
7  REGULAR  7626452  2609259  
8  REGULAR  7626464  2609272  
9  REGULAR  7626508  2609309

In [28]:
mta_data.DATE.sort_values().head(5)

3616662    05/01/2021
3564201    05/01/2021
3564202    05/01/2021
3564203    05/01/2021
3564204    05/01/2021
Name: DATE, dtype: object

In [29]:
mta_data.DATE.sort_values().tail(5)

21901    08/31/2021
21902    08/31/2021
21903    08/31/2021
72265    08/31/2021
71140    08/31/2021
Name: DATE, dtype: object

## Exploratory data analysis in pandas 

In [30]:
mta_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3679429 entries, 0 to 3679428
Data columns (total 11 columns):
 #   Column    Dtype 
---  ------    ----- 
 0   CA        object
 1   UNIT      object
 2   SCP       object
 3   STATION   object
 4   LINENAME  object
 5   DIVISION  object
 6   DATE      object
 7   TIME      object
 8   DESC      object
 9   ENTRIES   int64 
 10  EXITS     int64 
dtypes: int64(2), object(9)
memory usage: 308.8+ MB


In [31]:
# check column format
mta_data.columns

Index(['CA', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES', 'EXITS'],
      dtype='object')

In [32]:
mta_data.dtypes

CA          object
UNIT        object
SCP         object
STATION     object
LINENAME    object
DIVISION    object
DATE        object
TIME        object
DESC        object
ENTRIES      int64
EXITS        int64
dtype: object

In [33]:
#detect empty value
mta_data.isna().sum()

CA          0
UNIT        0
SCP         0
STATION     0
LINENAME    0
DIVISION    0
DATE        0
TIME        0
DESC        0
ENTRIES     0
EXITS       0
dtype: int64

In [34]:
import datetime

In [36]:
# Take the date and time fields into a single datetime column
mta_data["DATE_TIME"] = pd.to_datetime(mta_data.DATE+" "+ mta_data.TIME, format="%m/%d/%Y %H:%M:%S")

mta_data.head(5)

CA  UNIT       SCP STATION LINENAME DIVISION        DATE      TIME  \
0  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  00:00:00   
1  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  04:00:00   
2  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  08:00:00   
3  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  12:00:00   
4  A002  R051  02-00-00   59 ST  NQR456W      BMT  08/28/2021  16:00:00   

      DESC  ENTRIES    EXITS           DATE_TIME  
0  REGULAR  7625998  2609113 2021-08-28 00:00:00  
1  REGULAR  7626014  2609117 2021-08-28 04:00:00  
2  REGULAR  7626024  2609135 2021-08-28 08:00:00  
3  REGULAR  7626086  2609191 2021-08-28 12:00:00  
4  REGULAR  7626206  2609221 2021-08-28 16:00:00